lib/Monitor/Elembio/RunFolder.pm

```diff
@@ -427,30 +523,110 @@ sub _build__run_params_data {
+    return 1;
+  }
+  return 0;
+}
+
+=head2 is_completed
+
+When the run has been successfully completed return 1. Otherwise 0.
+
+=cut
+has q{is_completed}  => (
+  isa               => q{Bool},
+  is                => q{ro},
+  required          => 0,
+  lazy_build        => 1,
+);
+sub _build_is_completed {
+  my $self = shift;
+  my $upload_data = $self->_run_uploaded_data();
+  my $data_ok = defined $upload_data and exists $upload_data->{$OUTCOME};
+  if ( $data_ok and $upload_data->{$OUTCOME} eq $OUTCOME_COMPLETE) {
+    return 1;
+  }
+  return 0;
+}
+
+=head2 _run_uploaded_data
+
+Reference to hash object that represents the JSON file
+content of RunUploaded.json.
+
+=cut
+has q{_run_uploaded_data} => (
+  isa               => q{Maybe[HashRef]},
+  is                => q{ro},
+  required          => 0,
+  init_arg          => undef,
+  lazy_build        => 1,
+);
+sub _build__run_uploaded_data {
+  my $self = shift;
+  my $run_uploaded_file = catfile($self->runfolder_path, $RUN_UPLOAD_FILE);
+  my $json_data;
+  if ( -e $run_uploaded_file ) {
+    $json_data = decode_json(slurp $run_uploaded_file);
+  }
+  return $json_data;
+}
+
 =head2 process_run_parameters
 
 Core function of the class that is called on the run folder
 periodically to update dynamic properties of a run in the DB.
-Created and completed runs are updated accordingly with a status
-and a time stamp of the event in the DB.
+If a run record does not exist, it is created.
+The cycle count, tags and the following statuses are
+checked/assigned accordingly:
+- 'run in progress'   run basecalling is in progress
+- 'run stopped early' run finished with failed outcome
+- 'run complete'      run completed successfully
+Each of the above events saves a time stamp in the DB.
 
 =cut
 sub process_run_parameters {
   my $self = shift;
   my $run_row = $self->tracking_run();
   my $current_run_status_obj = $run_row->current_run_status;
   my $run_uploaded_path = catfile($self->runfolder_path, $RUN_UPLOAD_FILE);
-  my $is_run_complete = -e $run_uploaded_path;
   
   if ( ! $current_run_status_obj ) {
     $run_row->set_instrument_side($self->instrument_side, $USERNAME);
     $current_run_status_obj = $run_row->update_run_status($RUN_STATUS_INPROGRESS, $USERNAME, $self->date_created);
     $self->info('New run ' . $self->runfolder_path . ' created');
+    $self->_set_tags();
   }
 
   $self->_set_actual_cycle_count();
   
-  if ($is_run_complete) {
+  if ($self->is_failed) {
+    if ( $current_run_status_obj->run_status_dict->compare_to_status_description($RUN_STATUS_STOPPED) == -1 ) {
+      my $date = DateTime->from_epoch(epoch => (stat  $run_uploaded_path)[9]);
+      $run_row->update_run_status($RUN_STATUS_STOPPED, $USERNAME, $date);
+      $self->info('Run ' . $self->runfolder_path . ' is failed');
+    } else {
+      $self->info('Run ' . $self->runfolder_path . ' was failed, current status ' . $current_run_status_obj->description);
+    }
+  }
+
+  if ($self->is_completed) {
```

Comment:
add to the condition: or if the run is marked as failed, but the expected number of cycles was performed